### ISSUE TITLE
ipc_core: Native_socket_stream and Blob_stream_mq_*er sync_io APIs might close FD while use still allowed to operate on it (esp epoll).

### ISSUE THREAD

**Issue by ygoldfeld** [Flow-IPC/ipc_core#31]
[ CC @wkorbe ]

Mr. @wkorbe was, with my semi-help, debugging some strangeness he was seeing when doing ipc_session work. The strangeness turned out to be not an internal Flow-IPC problem but some other library code that was being naughty with FDs that didn't belong to it. However this activity also made me worry about a related thing (details follow) which could be a problem. And turns out that at least theoretically it could indeed be a problem.

To be clear, no code - whether the considerably amounts of functional and unit testing in this open-source project, nor Will's application - has as-yet suffered from this potential problem. Nevertheless:
- that doesn't mean it'll never happen; and
- supposing it never does happen, the problem still exists in the sense that it's an inelegant loose end that adds entropy and can cause some kind of unexpected pain in the future. It's better to be clean and correct!

### What's affected?

Only `sync_io::` APIs are affected; of those specifically only the following are affected:
- ipc::transport::sync_io::Native_socket_stream
- ipc::transport::sync_io::Blob_stream_mq_sender [with POSIX MQ only, not bipc MQ]
- Therefore, also, ipc::transport::struc::sync_io::Channel operating on top of the above is affected. (That's... all of them [`struc::sync_io::Channel`s] at the moment.)

Async-I/O APIs built around the above (basically, remove the `sync_io::` fragment) - and, in turn, anything built using them (including various session stuff) *is not affected*, because they use boost.asio-style async-wait implementations which, basically, won't touch the potentially-problematic FDs, once they are possibly closed. Only direct epoll (etc.) integrations would be (potentially) problematic.

### The problem

In retrospect it's pretty straightforward. Take API object sync_io::X that uses a socket/something S such that in high-performance fashion it needs to wait asynchronously on event "readable" (and the same holds for "writable") on S. In sync_io pattern, X will inform (synchronously) the user that indeed the user should async-wait for "readable" on S; and when it is to call-in to a particular X API to inform it of this fact. X internal code will then react to it appropriately.

It is assumed (though a to-do below states we should make this ultra-super-clear and perhaps emphasized in the docs, based on what I've now seen) the user will not touch S in any way after X::~X() or equivalent (replacement via operator= perhaps) begins execution.

Well-behaving code will also not touch S after X reports that X is hosed via an error-report (via callback or a send()-type function returning it synchronously), except to perhaps deregister it from an epoll-wait set. However, this is not a formal requirement, actually. In any case, though, even the deregistration-from-epoll-wait-set is still touching S. To be clear, deregistration-from-epoll-wait-set (whether or on error, or generally just before deleting X) is common-sense and good on the user's part.

The point: Sometimes internal Flow-IPC (ipc_core actually) code will in actual fact close() S, potentially long before X::~X() (or replacement via operator=). Now suppose the user tries to do something with S. Could be bad.

That's basically it, in short. The most direct fix is to "simply" not close() S until X::~X() (or equivalent, replacement via operator=).

### Actual danger (more detail)

Now let's discuss it in more detail. The problem is in the impls of Native_socket_stream and Blob_stream_mq_sender.

The X code that close()s S is, fortunately, reasonably well structured/not spaghetti code. There is - in both cases - a clause (2x in Native_socket_stream and 1x in Blob_stream_mq_sender [applicable with POSIX MQ only], though notably not in Blob_stream_mq_receiver) that, upon detecting a system error on the underlying socket (m_peer_socket, m_mq respectively), it nullifies the unique_ptr handle to that socket-storing object (`m_peer_socket.reset()`, `m_mq.reset` respectively). This will execute the destructor of the underlying object. In the case of m_peer_socket, that's a boost::asio::local::stream_protocol::socket destructor which is documented to close the socket; hence it does `::close()` on it ultimately.

Then, at the next opportunity, X reports the hosing of `*this` to X user.

Now let's discuss what user might do at this stage (that could be bad) specifically. If they're using boost.asio to perform async-waits, then there's no need or reason to touch S again. (E.g., our eating-own-dog-food async-I/O wrappers use boost.asio.) However consider code (such as what @wkorbe was doing) that uses native FD-watching. If using select() or poll() or equivalent, then there's no need to ever mention S again. However if using epoll or equivalent, user will likely want to deregister S from whatever epoll-set or equivalent. With epoll that's epoll_ctl(EPOLL_CTL_DEL) to remove S from epoll-set FD <whatever>.

But - S is in Linux an int, and it might now refer to *no* FD (EBAD) or even to a completely new FD. What will actually happen? Well, actually, possibly nothing terrible. If it's a bad FD, no harm done - assuming user code does not plotz on EBADF in this situation (not necessarily a correct assumption). It it now refers to some other file-description then... hard to say exactly. Actually, due to what [this link](https://idea.popcount.org/2017-03-20-epoll-is-fundamentally-broken-22/) calls an epoll mis-design, the FD might even still refer to an active file-description. It's hard to be sure for me at the moment without diving into the topic even more, maybe reading kernel... docs... you know.

EDIT: According to that link, and we have seen this in action, close()ing a thing that's being epoll_wait()ed upon causes the epoll_wait() to wake up and report an active event. Except the FD is no longer valid at that moment, and unless some race opens another FD onto that same int, what would happen is... the code might report it to the object X. At that point - since my code is defensive in that sense - it'll notice X is hosed already and won't proceed further. So, perhaps through luck nothing terrible would come to pass.

However, that's a digression. Plainly, it is bad and entropy-laden to kill an FD and then user attempts to unregister it from some other kernel object. We might just be lucky that that it happens to not be fatally terrible.

TL;DR: Native_socket_stream, Blob_stream_mq_sender (POSIX-MQ template arg only) detecting hosing of the socket/MQ, followed by user deregistering watched-FD via epoll_ctl(EPOLL_CTL_DEL) = potentially trying to operate on a close()d FD, formally undefined behavior though in practice *might* be harmless.

### How to fix

There are a couple approaches... but I'll skip the hemming/hawing about ones I've discarded. I'll just say one idea was to dup() the real FD and have user watch that one - but it's at best ugly/hard to maintain; and quite possibly wouldn't even really work. Details omitted/waste of effort.

Suggested fix (and discussion):

Basically, just don't nullify m_peer_socket / m_mq on error. Simply let destructor do it, by which point it is contractually safe and correct. (Plus that's what happens when there is no socket-hosing error detected.)

How to structure this fix without affecting a bunch of other things in the impl? Answer:

Well, why am I nullifying it in the first place? The comment in both locations basically says something like, "might as well return the resource to the OS." Well, we've now determined that is folly and is subsumed by the important desire to not cause chaos. So, whatever.

In that case... is the fact m_mq / m_peer_socket has been potentially nullified in some way checked by other code? Simply not-nullifying it would lead to changes in behavior (unintended side effects of fix), if so. Answer: Yes, actually, there are a few `if`s that do A or B depending on m_mq / m_peer_socket being truthy or falsy.

In Blob_stream_mq_sender it's an extremely straightforward single check, in dtor.

In Native_socket_stream the check is part of a quite sophisticated complex algorithm involving two quasi-independent mutually-facing pipes on one FD.

All in all, it is best not to disturb the existing algorithms. Where they check for "socket has been hosed," they should just keep checking for it. We should decouple this (essentially) "socket has been hosed" flag from the act of destroying the underlying socket-bearing object - which should be delayed until dtor.

The mechanics of this decoupling... there are various possibilities. Some of them are super-pretty but arguably overkill given the small magnitude of our task here. Whatever. Right now I favor the following straightforward trick:

- In both cases it's unique_ptr<Something> m_guy (m_peer_socket or m_mq).
- So keep that; but add a data member right after m_guy: unique_ptr<Something> m_hosed_guy. It starts null; is protected by same mutex (if any) as m_guy (in reality applicable only in N_s_s); and if it is non-null, then m_guy is null.
- Whenever the code does `m_guy.reset()`, we instead do: `m_hosed_guy = std::move(m_guy); assert(!m_guy);`.
- Other than that, m_hosed_guy is never mentioned or touched in any way.

Basically m_hosed_guy is m_guy on death row. When death comes for `*this` (dtor), then either m_guy or m_hosed_guy gets destroyed, while the other is null (or both are null). Simple. But! Do a grep-equivalent for m_mq and m_peer_socket: Make sure I have not missed some m_peer_socket-affecting code in the above analysis.

### Also (doc/comment work)

Relatedly, we should clarify things in documentation, particularly when it comes to sync_io-pattern users who use not-boost.asio (and likely not-select()/poll()/etc. but rather epoll() and similar). At the moment these are either assumed or not mentioned... in any case not explicitly stated.

- Make clear rules for sync_io-pattern users as to how they are *allowed* to operate on any FD they've been asked to async-wait-upon. Namely:
  - Remind user to not touch such an FD after dtor begins.
  - Require epoll_ctl(EPOLL_CTL_DEL) or equivalent (if _ADDed already) ahead of entering dtor. Suggest doing so at least then; but also perhaps upon being informed of object-being-hosed.
  - Informally, only read-only "watching" activities on the socket are allowed. (In particularly anything like close() leads to undefined behavior/actual mayhem. @wkorbe has observed this.)

### Priority

High. The next release must do this. We don't want entropy-laden stuff out there.

**Comment by ygoldfeld** [comment]
Before I forget: here's a WIP branch of ipc_core (and ipc, similarly-named): tkt31_sync_io_early_close_fd

Likely that's very close to final fix, but I haven't created a PR yet.
